src/librustc/traits/error_reporting.rs

```diff
@@ -640,16 +640,38 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                     ty::Predicate::ClosureKind(closure_def_id, kind) => {
                         let found_kind = self.closure_kind(closure_def_id).unwrap();
                         let closure_span = self.tcx.hir.span_if_local(closure_def_id).unwrap();
+                        let node_id = self.tcx.hir.as_local_node_id(closure_def_id).unwrap();
                         let mut err = struct_span_err!(
                             self.tcx.sess, closure_span, E0525,
                             "expected a closure that implements the `{}` trait, \
                                 but this closure only implements `{}`",
                             kind,
                             found_kind);
-                        err.span_note(
-                            obligation.cause.span,
-                            &format!("the requirement to implement \
-                                        `{}` derives from here", kind));
+
+                        let infer_tables = match self.tables {
+                            InferTables::Interned(tables) =>
+                                Some(InferTablesRef::Interned(tables)),
+                            InferTables::InProgress(tables) =>
+                                Some(InferTablesRef::InProgress(tables.borrow())),
+                            InferTables::Missing => None,
+                        };
+
+                        if let Some(tables) = infer_tables {
+                            if let Some(&(ty::ClosureKind::FnOnce, Some((span, name)))) =
```

Comment:
Good question. I think it'd be better to match more cases, and keep the message fairly tailored -- there are really only two cases. The closure might be `FnOnce` (in which case we can talk about something being moved) or it might be `FnMut` (in which case we can talk about something being written to).